doc/designs/app-passwords.md

```diff
@@ -0,0 +1,312 @@
+`simpleSecurityObject` into the `apppw` object.
+
+The LDIF code to add the `cn=apps,cn=accounts,$SUFFIX` entry, as well as new
+ACIs for access to the `apppw` object and its attributes, during
+* new installations will be in `install/share/apppw.ldif`,
+* upgrades will be in `install/updates/19-apppw.update` (entry) and
+  `install/updates/20-aci.update` (ACIs).
+
+There will be the following ACIs on the `cn=apps,cn=accounts,$SUFFIX` entry:
+```
+aci: (target = "cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (version 3.0; acl "Allow users to add entries grouping app passwords for a
+     particular app";
+     allow (add) 
+     userdn = "ldap:///all;)
+aci: (target = "ldap:///uid=($dn)-\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to add an app password for themselves";
+     allow (add) 
+     userdn = "ldap:///uid=($dn),cn=users,cn=accounts,$SUFFIX";)
+aci: (targetattr = "managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to search for their app passwords";
+     allow (search) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (targetattr = "displayName")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to read display names for their app
+     passwords";
+     allow (read) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+aci: (target = "ldap:///uid=\*,cn=\*,cn=apps,cn=accounts,$SUFFIX")
+     (targetattr = "uid || displayName || userPassword || appName || entryUUID
+     || managedBy")
+     (targetfilter=(objectClass=ipaAppPw))
+     (version 3.0; acl "Allow users to delete their app password";
+     allow (delete) 
+     userdn = "ldap:///self" OR
+     userattr = "managedBy#USERDN";)
+```
+
+Backup and restore is out of scope of this extension, but it will probably work
+with the new functionality.
+
+I do not need to enforce anything for app passwords. I suppose it will then be
+best to implement some if-else branching in
+`daemons/ipa-slapi-plugins/ipa-pwd-extop/prepost.c`, where it will be tested if
+the Password attribute being modified is not from the subtree for app passwords,
+and if it is, then no check and enforcements, as well as no OTP validation will
+be performed for it.
+
+**New dependencies**: TBD, not expected for now
+
+## Feature management
+
+The login process will remain the same, only instead of entering the primary
+password, the user will enter an app password. And as already stated, the user
+then will not have access to the account management.
+
+### UI
+
+TBD
+
+### CLI
+
+| Command | Arguments | Options |
+| ------- | --------- | ------- |
+| apppw-add | username (user's uid), displayname, appname |  |
+| apppw-list | username (user's uid) |  |
+| apppw-del | username (user's uid), uuid |  |
```

Comment:
The IPA CLI CRUD commands are -add, -del, -mod, -show and -find. I suspect list == find in this context. I suppose there is no value in mod unless you want to allow the displayname to be changed. I guess -show isn't all that interesting but it might be required by the UI.

As already mentioned, there needs to be a separate CLI for managing the appnames.

Reply:
As has already been noted, appnames will be possibly an attribute of app passwords, definitely not a separate mechanism.

As for the commands:
* `-mod` would indeed be redundant - I came to a conclusion that it is not worth it to introduce it just for changing the display name of an app password, and it is even questionable whether such possibility would be desirable.
* `-list` is indeed basically an analogy of `-find`, but with one important restriction: currently, I plan to always return all user's passwords using the list command. It would even be optimal if this command would not take any argument as a search filter, but as I will probably have to inherit from `LDAPSearch`, I may end up always overriding a search filter with someting like `*` (and possibly writing out a warning in the output that this was done). This is being discussed in the point 9 of [this post above](https://github.com/freeipa/freeipa/pull/4061#issuecomment-638803574) and subsequent answer, and is still unresolved.
* As for `-show`, I thought it is undesirable, as there must not be an ability to retrieve an app password itself, but it is true that it may be used for retrieving the display name and UUID to show in the output - both in CLI and in Web UI. The question is: should this rather be a part of `-list`, or should this be in a non-public, internal `-show` command that will be internally called by `-list` for displaying an app password identification? It still seems to me that it would be better for this to be a direct part of the `-list` command.

Reply:
Ok with -mod but we may get an RFE on it eventually

You must use -find and not -list to meet the API pattern.

Right, -show would just show the displayname you requested so I doubt the utility of it. I don't know if the UI depends on this being there though.

Reply:
> Ok with -mod but we may get an RFE on it eventually

It will not be so hard to add it if it will be requested. But if you think that such feature really is desirable, I can implement it right away.

> You must use -find and not -list to meet the API pattern.

OK, so it will be `apppw-find`.

> Right, -show would just show the displayname you requested so I doubt the utility of it. I don't know if the UI depends on this being there though.

This needs to be explored. I elaborated on the UI part of the implementation just during the weekend, and I don't exactly know how will the implementation of the list of app passwords look like. If `apppw-show` will be needed for this, then I should implement it as well.